lib/lf/fe/loc_comp_ellbvp.h

```diff
@@ -20,214 +20,236 @@
-                    std::cout << ref_el << ", shape = \n"
-                              << (geo_ptr->Global(ref_el_corners))
-                              << std::endl);
-  // Midpoints of edges in the reference cell
-  Eigen::MatrixXd ref_mp(2, 4);
-  switch (ref_el) {
-    case lf::base::RefEl::kTria(): {
-      // clang-format off
-       ref_mp << 0.5,0.5,0.0,-1.0,
-                 0.0,0.5,0.5,-1.0;
-      // clang-format on
-      break;
-    }
-    case lf::base::RefEl::kQuad(): {
-      // clang-format off
-       ref_mp << 0.5,1,0.5,0.0,
-                 0.0,0.5,1.0,0.5;
-     // clang format on
-     break;
-   }
-   default: {
-     LF_ASSERT_MSG(false,"Illegal entity type!");
-     break;
-   }
-   } //end switch
-   // Midpoints of edges in world coordinates
-   const Eigen::MatrixXd mp(geo_ptr->Global(ref_mp));
-
-   const double area = lf::geometry::Volume(*geo_ptr);
-
-   // Vector for returning element vector
-   ElemVec elem_vec = ElemVec::Zero();
-   // Run over the midpoints of edges and fetch values of the source function
-   // there
-  for (int k = 0; k < num_nodes; k++) {
-     const auto fval_half = 0.5*f_(mp.col(k));
-     elem_vec[k] += fval_half;
-     elem_vec[(k+1)%num_nodes] += fval_half;
-   }
-   SWITCHEDSTATEMENT(
-       dbg_ctrl, dbg_locvec,
-       std::cout << "element vector = " <<
-       elem_vec.head(num_nodes).transpose()
-                 << std::endl);
-   return (area/num_nodes)*elem_vec;
- }
-
 /**
  * @brief Class for local quadrature based computations for Lagrangian finite
- * elements
+ * elements and second-order scalar elliptic BVPs.
  *
+ * @tparam SCALAR type for the entries of the element matrices
+ * @tparam DIFF_COEFF a functor providing point evaluation for the diffusion
+ * tensor
+ * @tparam REACTION_COEFF a functor for point evaluation of the reaction
+ * coefficient
  *
+ * ### Template parameter type requirements
+ *
+ * - SCALAR must be a field type like `std::complex<double>`
+ * - DIFF_COEFF must comply with `std::function<T(Eigen::Vector2d)>`, where `T`
+ * is either a SCALAR compatible type of a matrix type like
+ * `Eigen::Matrix<SCALAR,...>`.
+ * - REACTION_COEFF must behave like `std::function<SCALAR(Eigen::Vector2d)>`.
+ *
+ * @note This class complies with the type requirements for the template
+ argument
+ * ELEM_MAT_COMP of the function lf::assemble::AssembleMatrixLocally().
+ *
+ * The element matrix is that for the (local) bilinear form
+ * @f[
+    (u,v) \mapsto\int\limits_{K}\boldsymbol{\alpha}(\mathbf{x})\mathbf{grad}\,u
+          \cdot\mathbf{grad}\,v + \gamma(\mathbf{x})u\,v\,\mathrm{d}\mathbf{x}
+ \;,
+ * @f]
+ * with _diffusion coefficient_ @f$\mathbf{\alpha}@f$ and reaction coefficient
+ @f$\gamma@f$.
  */
-template <typename DIFF_COEFF, typename REACTION_COEFF>
-class LagrangeFEEllBVPElementMatrix {
+template <typename SCALAR, typename DIFF_COEFF, typename REACTION_COEFF>
+class LagrangeFEEllBVPElementMatrix : public LocCompLagrFEPreprocessor {
```

Comment:
Please, don't inherit behavior from `LocCompLagrFEPreprocessor`. The statement "every `LagrangeFEEllBvpElementMatrix` is a `LocCompLagrFEPreprocessor`" doesn't make any sense. I would store a reference to a `LocCompLagrFEPreprocessor` in a membervariable of `LagrangeFEEllBVPElementMatrix`.

Reply:
Also, why does this class only work for Lagrange-type Finite elements?

Reply:
Oh and finally, I would split this class up into two classes:
- one class represents the bilinear form (u,v) \mapsto\int\limits_{K}\boldsymbol{\alpha}(\mathbf{x})\mathbf{grad}\,u
          \cdot\mathbf{grad}\,v
- the other represents the bilinear form (u,v) \mapsto \gamma(\mathbf{x})u\,v\,\mathrm{d}\mathbf{x}

Also it would be important that both classes implement a common interface, e.g. `BilinearForm`.

In order to support the addition of two bilinear forms, we can write a third class that also implements the `BilinearForm` interface and that takes two other `BilinearForm`'s and adds them. This is the well-known "composite-pattern" with which I've made excellent experiences.